File: strategy/sizing.py
from __future__ import annotations

import logging
from typing import Optional

from core.config import PositionConfig
from core.models import AccountState, TradeSignal
from data_sources.market_data import MarketData

logger = logging.getLogger(__name__)


class PositionSizer:
    def __init__(self, config: PositionConfig, market_data: MarketData) -> None:
        self._cfg = config
        self._md = market_data
        self._consecutive_losses: int = 0
        # 런타임 오버라이드: 텔레그램 /capital 명령으로 설정 가능.
        # 설정되면 실제 잔고 무시하고 이 금액(USDC)을 마진 기준으로 사용.
        # None이면 잔고 × base_capital_pct 자동.
        self._fixed_capital: Optional[float] = None

    def notify_loss(self) -> None:
        self._consecutive_losses += 1

    def notify_win(self) -> None:
        self._consecutive_losses = 0

    def set_fixed_capital(self, amount: Optional[float]) -> None:
        """봇 실행 중 자본 금액 변경. None이면 자동(잔고 비율)로 복원."""
        self._fixed_capital = amount
        if amount is None:
            logger.info("자본 모드 → 자동(잔고 × base_capital_pct)")
        else:
            logger.info("자본 모드 → 고정 $%.2f USDC", amount)

    def get_capital_mode(self) -> str:
        if self._fixed_capital is None:
            return f"AUTO (잔고 × {self._cfg.base_capital_pct*100:.0f}%)"
        return f"FIXED ${self._fixed_capital:.2f}"

    def compute_quantity(
        self,
        signal: TradeSignal,
        account: AccountState,
        drawdown_reduction_pct: float = 0.0,
    ) -> float:
        """
        주문 수량(코인 단위) 계산.
        drawdown_reduction_pct: 0이면 정상, 0.5이면 절반 사이즈
        """
        if self._fixed_capital is not None:
            # 고정 금액 모드: 잔고 무시
            base_usdt = self._fixed_capital
        else:
            base_usdt = account.balance_usdt * self._cfg.base_capital_pct

        # 드로우다운 축소
        if drawdown_reduction_pct > 0:
            base_usdt *= (1.0 - drawdown_reduction_pct)
            logger.info("드로우다운 사이즈 축소: %.0f%%", drawdown_reduction_pct * 100)

        # 연속 손실 축소
        if self._consecutive_losses >= 3:
            base_usdt *= 0.5
            logger.info("연속 손실 %d회 — 사이즈 50%% 축소", self._consecutive_losses)

        # 변동성 조정
        try:
            target_atr = self._md.compute_atr(signal.symbol, period=14)
            avg_atr = self._md.compute_atr(signal.symbol, period=50)
            if avg_atr > 0 and target_atr > 0:
                vol_adj = avg_atr / target_atr
                # 0.5~1.5 범위 클램프
                vol_adj = max(0.5, min(1.5, vol_adj))
                base_usdt *= vol_adj
                logger.debug("변동성 조정 계수: %.3f", vol_adj)
        except Exception as e:
            logger.warning("변동성 조정 실패: %s", e)

        # 최대 포지션 자본 제한 (FIXED 모드에선 사용자가 명시한 금액 그대로 사용)
        if self._fixed_capital is not None:
            notional_usdt = base_usdt
        else:
            max_usdt = account.balance_usdt * self._cfg.max_position_capital_pct
            notional_usdt = min(base_usdt, max_usdt)

        # 레버리지 적용 — notional = quantity * price
        price = signal.avg_entry_price
        if price <= 0:
            logger.error("진입가 0 — 수량 계산 불가")
            return 0.0

        # 증거금 기준 notional 역산
        margin_usdt = notional_usdt
        notional_with_lev = margin_usdt * self._cfg.leverage
        quantity = notional_with_lev / price

        logger.info(
            "사이즈 계산: 마진 $%.2f × %dx레버 / $%.4f = %.6f %s",
            margin_usdt, self._cfg.leverage, price, quantity, signal.symbol,
        )
        return round(quantity, 6)

File: strategy/sizing.rs
use std::sync::{Arc, RwLock};
use tracing::{info, warn, debug};
use crate::core::config::PositionConfig;
use crate::core::models::{AccountState, TradeSignal};
use crate::data_sources::market_data::MarketData;

pub struct PositionSizer {
    config: PositionConfig,
    market_data: Arc<MarketData>,
    consecutive_losses: RwLock<i32>,
    fixed_capital: RwLock<Option<f64>>,
}

impl PositionSizer {
    pub fn new(config: PositionConfig, market_data: Arc<MarketData>) -> Self {
        Self {
            config,
            market_data,
            consecutive_losses: RwLock::new(0),
            fixed_capital: RwLock::new(None),
        }
    }

    pub fn notify_loss(&self) {
        let mut losses = self.consecutive_losses.write().unwrap();
        *losses += 1;
    }

    pub fn notify_win(&self) {
        let mut losses = self.consecutive_losses.write().unwrap();
        *losses = 0;
    }

    pub fn set_fixed_capital(&self, amount: Option<f64>) {
        let mut fixed = self.fixed_capital.write().unwrap();
        *fixed = amount;
        if let Some(amt) = amount {
            info!("Capital mode -> FIXED ${:.2} USDC", amt);
        } else {
            info!("Capital mode -> AUTO (Balance x base_capital_pct)");
        }
    }

    pub fn get_capital_mode(&self) -> String {
        let fixed = self.fixed_capital.read().unwrap();
        if let Some(amt) = *fixed {
            format!("FIXED ${:.2}", amt)
        } else {
            format!("AUTO (Balance x {:.0}%)", self.config.base_capital_pct * 100.0)
        }
    }

    pub async fn compute_quantity(
        &self,
        signal: &TradeSignal,
        account: &AccountState,
        drawdown_reduction_pct: f64,
    ) -> f64 {
        let mut base_usdt = {
            let fixed = self.fixed_capital.read().unwrap();
            if let Some(amt) = *fixed {
                amt
            } else {
                account.balance_usdt * self.config.base_capital_pct
            }
        };

        if drawdown_reduction_pct > 0.0 {
            base_usdt *= 1.0 - drawdown_reduction_pct;
            info!("Drawdown size reduction: {:.0}%", drawdown_reduction_pct * 100.0);
        }

        let losses = *self.consecutive_losses.read().unwrap();
        if losses >= 3 {
            base_usdt *= 0.5;
            info!("Consecutive losses {} - size 50% reduction", losses);
        }

        // Volatility adjustment
        let target_atr = self.market_data.compute_atr(&signal.symbol, 14).await;
        let avg_atr = self.market_data.compute_atr(&signal.symbol, 50).await;
        if avg_atr > 0.0 && target_atr > 0.0 {
            let vol_adj = (avg_atr / target_atr).clamp(0.5, 1.5);
            base_usdt *= vol_adj;
            debug!("Volatility adjustment factor: {:.3}", vol_adj);
        }

        let notional_usdt = {
            let fixed = self.fixed_capital.read().unwrap();
            if fixed.is_some() {
                base_usdt
            } else {
                let max_usdt = account.balance_usdt * self.config.max_position_capital_pct;
                f64::min(base_usdt, max_usdt)
            }
        };

        let price = signal.avg_entry_price;
        if price <= 0.0 {
            warn!("Entry price 0 - cannot compute quantity");
            return 0.0;
        }

        let margin_usdt = notional_usdt;
        let notional_with_lev = margin_usdt * self.config.leverage as f64;
        let quantity = notional_with_lev / price;

        info!("Size calculation: Margin ${:.2} x {}x lev / ${:.4} = {:.6} {}",
              margin_usdt, self.config.leverage, price, quantity, signal.symbol);
        
        (quantity * 1_000_000.0).round() / 1_000_000.0
    }
}
